---
title: "PMBB Phenome-wide Association Studies"
output: html_document
editor_options: 
  chunk_output_type: console
---

<!-- Run this 'development' chunk -->
<!-- Store every call to library() that you need to explore your functions -->

```{r development, include=FALSE}
library(testthat)
```

<!--
 You need to run the 'description' chunk in the '0-dev_history.Rmd' file before continuing your code there.

If it is the first time you use {fusen}, after 'description', you can directly run the last chunk of the present file with inflate() inside.
--> 

```{r development-load}
# Load already included functions if relevant
pkgload::load_all(export_all = FALSE)
```

<!-- 
 Store your dataset in a directory named "inst/" at the root of your project.
 Use it for your tests in this Rmd thanks to `pkgload::load_all()` to make it available
and `system.file()` to read it in your examples.

- There already is a dataset in the "inst/" directory to be used in the examples below
-->

```{r development-dataset, eval=FALSE}
# Run all this chunk in the console directly
# There already is a dataset in the "inst/" directory
# Make the dataset file available to the current Rmd during development
pkgload::load_all(path = here::here(), export_all = FALSE)
# 
# # You will be able to read your example data file in each of your function examples and tests as follows - see chunks below
# datafile <- system.file("nyc_squirrels_sample.csv", package = "pmbbPheWASr")
# nyc_squirrels <- read.csv(datafile, encoding = "UTF-8")
# 
# nyc_squirrels
```


# Filter Variant Annotations File

The `pmbb_filter_exome_annotations` function filters the variant annotations file to only include variants within the specified gene. Because the variant annotation file is extremely large, this function wraps the `awk` shell command.
    
```{r function-pmbb_filter_exome_annotations_gene}
#' Filter the PMBB whole exome sequence annotations file
#' 
#' This function filters the variant annotations file to only include variants within the specified gene. Because the variant annotation file is extremely large, this function wraps the `awk` shell command
#' 
#' @param annotations_filename Path to the PMBB whole-exome sequencing variant annotations file
#' @param gene_names A character vector of gene names used to filter the variant annotations file 
#'
#' @return
#' A [tibble::tibble()] containing the filtered variant annotations 
#' 
#' 
#' @export

pmbb_filter_exome_annotations_gene <- function(annotations_filename, gene_names) {
  
  cli::cli_progress_step("Extracting {gene_names} from {.file {annotations_filename}}")
  
  exome_gene_annotations_filtered <- processx::run("awk",
    args = c(paste0("NR == 1 || ", paste0("($8 == \"", gene_names, "\")", collapse = " || ")), annotations_filename),
    echo = FALSE,
    echo_cmd = FALSE
  )
  
  exome_gene_annotations_df <- data.table::fread(text = exome_gene_annotations_filtered$stdout) %>%
    tibble::as_tibble()
  
  return(exome_gene_annotations_df)
}
```
  
```{r example-pmbb_filter_exome_annotations_gene}
pmbb_filter_exome_annotations_gene(annotations_filename = "/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt",
                              gene_names = c("LDLR"))
```
  
```{r tests-pmbb_filter_exome_annotations_gene}
test_that("pmbb_filter_exome_annotations works", {
  
  res <- pmbb_filter_exome_annotations_gene(annotations_filename = "/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt",
                              gene_names = c("LDLR"))
  
  expect_true(inherits(pmbb_filter_exome_annotations_gene, "function"))
  expect_true(tibble::is_tibble(res))
})
```



<!--
# There can be development actions

Create a chunk with 'development' actions

- The chunk needs to be named `development` or `dev`
- It contains functions that are used for package development only
- Note that you may want to store most of these functions in the 0-dev_history.Rmd file

These are only included in the present flat template file, their content will not be part of the package anywhere else.
-->

```{r development-inflate, eval=FALSE}
# Keep eval=FALSE to avoid infinite loop in case you hit the knit button
# Execute in the console directly
fusen::inflate(flat_file = "dev/flat_pmbb_phewas.Rmd", vignette_name = "Get started")
```

