---
title: "Get started"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{get-started}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(pmbbPheWASr)
```

<!-- WARNING - This vignette is generated by {fusen} from dev/flat_pmbb_phewas.Rmd: do not edit by hand -->

<!-- Run this 'development' chunk -->
<!-- Store every call to library() that you need to explore your functions -->


<!--
 You need to run the 'description' chunk in the '0-dev_history.Rmd' file before continuing your code there.

If it is the first time you use {fusen}, after 'description', you can directly run the last chunk of the present file with inflate() inside.
--> 


<!-- 
 Store your dataset in a directory named "inst/" at the root of your project.
 Use it for your tests in this Rmd thanks to `pkgload::load_all()` to make it available
and `system.file()` to read it in your examples.

- There already is a dataset in the "inst/" directory to be used in the examples below
-->


# Filter Variant Annotations File

The `awk_str_filter` function can be used to filter arbitrary rows from a tab-delimited text file using `awk`. The function accepts the path to a tab-delimited text file, name of a column to filter, and the string to filter for. The function returns a tibble containing the filtered rows. This function can be used to filter the PMBB variant annotations file to only include variants within the specified gene(s).
    

  

```{r example-awk_str_filter, eval = FALSE}
#' \dontrun{
ldlr_variants <- awk_str_filter(
  filename = "/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt",
  filter_col = "Gene.refGene",
  filter_str = c("LDLR"))
#' }
```

  

# Apply variant masks to annotation file

    
The `apply_variant_masks` function applies variant mask criteria to a data frame containing variant annotations. The function can accept multiple sets of filter criteria, each defined as a list of column names and filter conditions. The resulting filtered data frames are returned as a named list.


  

```{r example-apply_variant_masks, eval = FALSE}
#' \dontrun{
apply_variant_masks(
  ldlr_variants,
  masks = list(
    plof_0.001 = list(ExonicFunc.ensGene = "== 'stopgain'", gnomAD_exome_ALL = "< 0.001"),
    common_0.01 = list(gnomAD_exome_ALL = "> 0.01"),
    exonic_0.001 = list(Func.refGene = "== 'exonic'", gnomAD_exome_ALL = "< 0.001")
  )
)
#' }
```

  

  

```{r example-filter_dataframe, eval = FALSE, include = FALSE}
#' \dontrun{
filter_dataframe(
  ldlr_variants, 
  filter_list = list(ExonicFunc.ensGene = "== 'stopgain'", gnomAD_exome_ALL = "< 0.001"))
#' }
```

  

# Extracting genotypes

## Extract a single set of variants

The `pmbb_extract_genotypes` function is a wrapper around `plink2` to extract genotypes from PMBB plink files. The function accepts the path to the plink files and the list of variant IDs to extract. The function returns a long-format tibble containing the extracted genotypes for each PMBB participant.


  

```{r example-pmbb_extract_genotypes, eval = FALSE}
#' \dontrun{
pmbb_extract_genotypes(
  variant_df = ldlr_variants,
  variant_id_col = ID,
  effect_allele_col = Alt,
  plink_bin = "/project/voltron/Applications/PLINK/plink2_linux_avx2_20230607/plink2",
  bfile = "/project/PMBB/PMBB-Release-2020-2.0/Exome/pVCF/all_variants/PMBB-Release-2020-2.0_genetic_exome_GL"
)
#' }
```

  

## Extract genotypes based on masks 

The `pmbb_extract_genotype_masks` function extracts genotypes for a subset of participants based on a set of masks. This function wraps `awk_str_filter`, `apply_variant_masks`, and `pmbb_extract_genotypes` to extract genotypes for PMBB participants based on a set of masks. 
    


  

```{r example-pmbb_extract_genotype_masks, eval = FALSE}
#' \dontrun{
ldlr_mask_res <- pmbb_extract_genotype_masks(
  gene = "LDLR",
  annotation_file = "/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt",
  gene_col = "Gene.refGene",
  masks = list(
    plof_0.001 = list(ExonicFunc.ensGene = "== 'stopgain'", gnomAD_exome_ALL = "< 0.001"),
    rare_0.000001 = list(gnomAD_exome_ALL = "< 0.000001")
  ),
  mask_operator = list(
    plof_0.001 = "burden",
    rare_0.000001 = "single"
  ),
  variant_id_col = ID,
  effect_allele_col = Alt,
  plink_bin = "/project/voltron/Applications/PLINK/plink2_linux_avx2_20230607/plink2",
  bfile = "/project/PMBB/PMBB-Release-2020-2.0/Exome/pVCF/all_variants/PMBB-Release-2020-2.0_genetic_exome_GL")
#' }
```

  

# PheWAS convenience functions

## Format PMBB phecode data

The `pmbb_format_phecodes` function takes a path to a PMBB phecode file and applies a case threshold to categorize individuals as cases or controls. The function returns a tibble with `PMBB_ID` and phecodes as columns, with TRUE/FALSE values for each phecode column.
    

  

```{r example-pmbb_format_phecodes, eval = FALSE}
#' \dontrun{
pmbb_format_phecodes(phecode_file = "/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_phenotype_PheCode-matrix.txt")
#' }

```

  

  

    

  

```{r example-pmbb_format_covariates, eval = FALSE, include = FALSE}
#' \dontrun{
pmbb_format_covariates(
  covariate_files = c("/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_covariates.txt", "/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.1/PMBB-Release-2020-2.1_phenotype_covariates.txt"),
  populations = c("ALL", "EUR"),
  covariate_population_col = Class,
  covariate_cols = c(Age = Age_at_Enrollment, Sex = Gen_Sex, dplyr::starts_with("Genotype_PC"))
)
#' }

```

  

  


## Run a PheWAS

The `run_pmbb_phewas` function runs a PheWAS on the PMBB data, using the results of the `pmbb_extract_genotype_masks` function and paths to PMBB files containing phenotype and covariate data. 
    

  

```{r example-run_pmbb_phewas, eval = FALSE}
#' \dontrun{
# Extract genotypes for a series of masks
ldlr_mask_res <- pmbb_extract_genotype_masks(
  gene = "LDLR",
  annotation_file = "/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt",
  gene_col = "Gene.refGene",
  masks = list(
    plof_0.001 = list(ExonicFunc.ensGene = "== 'stopgain'", gnomAD_exome_ALL = "< 0.001"),
    common = list(gnomAD_exome_ALL = "> 0.01")
  ),
  mask_operator = list(
    plof_0.001 = "burden",
    common = "single"
  ),
  variant_id_col = ID,
  effect_allele_col = Alt,
  plink_bin = "/project/voltron/Applications/PLINK/plink2_linux_avx2_20230607/plink2",
  bfile = "/project/PMBB/PMBB-Release-2020-2.0/Exome/pVCF/all_variants/PMBB-Release-2020-2.0_genetic_exome_GL")

# Run a PheWAS
phewas_res <- run_pmbb_phewas(
  mask_genotypes_list = ldlr_mask_res,
  phecode_file = "/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_phenotype_PheCode-matrix.txt",
  covariate_files = c("/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_covariates.txt", "/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.1/PMBB-Release-2020-2.1_phenotype_covariates.txt"),
  populations = c("ALL", "EUR"),
  covariate_population_col = Class,
  covariate_cols = c(Age = Age_at_Enrollment, Sex = Gen_Sex, dplyr::starts_with("Genotype_PC"))
)
#' }
```

  

<!--

# There can be development actions

Create a chunk with 'development' actions

- The chunk needs to be named `development` or `dev`
- It contains functions that are used for package development only
- Note that you may want to store most of these functions in the 0-dev_history.Rmd file

These are only included in the present flat template file, their content will not be part of the package anywhere else.
-->


