---
title: "Get started"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{get-started}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(pmbbPheWASr)
```

<!-- WARNING - This vignette is generated by {fusen} from dev/flat_pmbb_phewas.Rmd: do not edit by hand -->

<!-- Run this 'development' chunk -->
<!-- Store every call to library() that you need to explore your functions -->


<!--
 You need to run the 'description' chunk in the '0-dev_history.Rmd' file before continuing your code there.

If it is the first time you use {fusen}, after 'description', you can directly run the last chunk of the present file with inflate() inside.
--> 


<!-- 
 Store your dataset in a directory named "inst/" at the root of your project.
 Use it for your tests in this Rmd thanks to `pkgload::load_all()` to make it available
and `system.file()` to read it in your examples.

- There already is a dataset in the "inst/" directory to be used in the examples below
-->


# Filter Variant Annotations File

The `awk_str_filter` function can be used to filter arbitrary rows from a tab-delimited text file using `awk`. The function accepts the path to a tab-delimited text file, name of a column to filter, and the string to filter for. The function returns a tibble containing the filtered rows. This function can be used to filter the PMBB variant annotations file to only include variants within the specified gene(s).
    

  

```{r example-awk_str_filter, eval = FALSE}
#' \dontrun{
ldlr_variants <- awk_str_filter(
  filename = "/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt",
  filter_col = "Gene.refGene",
  filter_str = c("LDLR"))
#' }
```

  

# Apply variant masks to annotation file

    
The `apply_variant_masks` function applies variant mask criteria to a data frame containing variant annotations. The function can accept multiple sets of filter criteria, each defined as a list of column names and filter conditions. The resulting filtered data frames are returned as a named list.


  

```{r example-apply_variant_masks, eval = FALSE}
#' \dontrun{

# Extract variants in a gene
ldlr_variants <- awk_str_filter(
  filename = "/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt",
  filter_col = "Gene.refGene",
  filter_str = c("LDLR"))

# Apply the specified masks to the set of variants
apply_variant_masks(
  ldlr_variants,
  masks = list(
    plof_0.001 = list(ExonicFunc.ensGene = "== 'stopgain'", gnomAD_exome_ALL = "< 0.001"),
    common_0.01 = list(gnomAD_exome_ALL = "> 0.01"),
    exonic_0.001 = list(Func.refGene = "== 'exonic'", gnomAD_exome_ALL = "< 0.001")
  )
)
#' }
```

  

  

```{r example-filter_dataframe, eval = FALSE, include = FALSE}
#' \dontrun{
filter_dataframe(
  ldlr_variants, 
  filter_list = list(ExonicFunc.ensGene = "== 'stopgain'", gnomAD_exome_ALL = "< 0.001"))
#' }
```

  

# Extracting genotypes

## Extract a single set of variants

The `pmbb_extract_genotypes` function is a wrapper around `plink2` to extract genotypes from PMBB plink files. The function accepts the path to the plink files and the list of variant IDs to extract. The function returns a long-format tibble containing the extracted genotypes for each PMBB participant.


  

```{r example-pmbb_extract_genotypes, eval = FALSE}
#' \dontrun{
pmbb_extract_genotypes(
  variant_df = ldlr_variants,
  variant_id_col = ID,
  effect_allele_col = Alt,
  plink_bin = "/project/voltron/Applications/PLINK/plink2_linux_avx2_20230607/plink2",
  bfile = "/project/PMBB/PMBB-Release-2020-2.0/Exome/pVCF/all_variants/PMBB-Release-2020-2.0_genetic_exome_GL"
)
#' }
```

  

## Extract genotypes based on masks 

The `pmbb_extract_genotype_masks` function extracts genotypes for a subset of participants based on a set of masks. This function wraps `awk_str_filter`, `apply_variant_masks`, and `pmbb_extract_genotypes` to extract genotypes for PMBB participants based on a set of masks. 
    


  

```{r example-pmbb_extract_genotype_masks, eval = FALSE}
#' \dontrun{
ldlr_mask_res <- pmbb_extract_genotype_masks(
  gene = "LDLR",
  annotation_file = "/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt",
  gene_col = "Gene.refGene",
  masks = list(
    plof_0.001 = list(ExonicFunc.ensGene = "== 'stopgain'", gnomAD_exome_ALL = "< 0.001"),
    rare_0.000001 = list(gnomAD_exome_ALL = "< 0.000001")
  ),
  mask_operator = list(
    plof_0.001 = "burden",
    rare_0.000001 = "single"
  ),
  variant_id_col = ID,
  effect_allele_col = Alt,
  plink_bin = "/project/voltron/Applications/PLINK/plink2_linux_avx2_20230607/plink2",
  bfile = "/project/PMBB/PMBB-Release-2020-2.0/Exome/pVCF/all_variants/PMBB-Release-2020-2.0_genetic_exome_GL")
#' }
```

  

# PheWAS convenience functions

## Format PMBB phecode data

The `pmbb_format_phecodes` function takes a path to a PMBB phecode file and applies a case threshold to categorize individuals as cases or controls. The function returns a tibble with `PMBB_ID` and phecodes as columns, with TRUE/FALSE values for each phecode column.
    

  

```{r example-pmbb_format_phecodes, eval = FALSE}
#' \dontrun{
pmbb_format_phecodes(phecode_file = "/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_phenotype_PheCode-matrix.txt")
#' }

```

  

  

    

  

```{r example-pmbb_format_covariates, eval = FALSE, include = FALSE}
#' \dontrun{
pmbb_format_covariates(
  covariate_files = c("/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_covariates.txt", "/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.1/PMBB-Release-2020-2.1_phenotype_covariates.txt"),
  populations = c("ALL", "EUR"),
  covariate_population_col = Class,
  covariate_cols = c(Age = Age_at_Enrollment, Sex = Gen_Sex, dplyr::starts_with("Genotype_PC"))
)
#' }

```

  

  


## Run a PheWAS

The `run_pmbb_phewas` function runs a PheWAS on the PMBB data, using the results of the `pmbb_extract_genotype_masks` function and paths to PMBB files containing phenotype and covariate data. 
    

  

```{r example-run_pmbb_phewas, eval = FALSE}
#' \dontrun{
# Extract genotypes for a series of masks
ldlr_mask_res <- pmbb_extract_genotype_masks(
  gene = "LDLR",
  annotation_file = "/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt",
  gene_col = "Gene.refGene",
  masks = list(
    plof_0.001 = list(ExonicFunc.ensGene = "== 'stopgain'", gnomAD_exome_ALL = "< 0.001"),
    common = list(gnomAD_exome_ALL = "> 0.01")
  ),
  mask_operator = list(
    plof_0.001 = "burden",
    common = "single"
  ),
  variant_id_col = ID,
  effect_allele_col = Alt,
  plink_bin = "/project/voltron/Applications/PLINK/plink2_linux_avx2_20230607/plink2",
  bfile = "/project/PMBB/PMBB-Release-2020-2.0/Exome/pVCF/all_variants/PMBB-Release-2020-2.0_genetic_exome_GL")

# Run a PheWAS
phewas_res <- run_pmbb_phewas(
  mask_genotypes_list = ldlr_mask_res,
  phecode_file = "/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_phenotype_PheCode-matrix.txt",
  covariate_files = c("/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_covariates.txt", "/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.1/PMBB-Release-2020-2.1_phenotype_covariates.txt"),
  populations = c("ALL", "EUR"),
  covariate_population_col = Class,
  covariate_cols = c(Age = Age_at_Enrollment, Sex = Gen_Sex, dplyr::starts_with("Genotype_PC"))
)
#' }
```

  

## Run a PheWAS and generate a report

The `run_pmbb_phewas_report` function is a wrapper around the `run_pmbb_phewas` function that generates an HTML report of the results.


  

```{r example-run_pmbb_phewas_report, eval = FALSE}
#' \dontrun{
run_pmbb_phewas_report(
  gene = c("LDLR"),
  annotation_file = "/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt",
  gene_col = "Gene.refGene",
  masks = list(
    plof_0.01 = list(ExonicFunc.ensGene = "%in% c('stopgain', 'stoploss', 'frameshift substitution')", gnomAD_exome_ALL = "< 0.01"),
    missense_0.01 = list(ExonicFunc.ensGene = "%in% c('nonsynonymous SNV')", gnomAD_exome_ALL = "< 0.01")
  ),
  mask_operator = list(
    plof_0.01 = "burden",
    missense_0.01 = "burden"
  ),
  variant_id_col = ID,
  effect_allele_col = Alt,
  populations = c("ALL"),
  covariate_population_col = "Class",
  covariate_cols = c(Age = Age_at_Enrollment, Sex = Gen_Sex, dplyr::starts_with("Genotype_PC")),
  mask_output_path = "LDLR_mask_results.rds",  # Optional: Specify the output path for the mask results file
  phewas_output_path = "LDLR_phewas_results.rds",  # Optional: Specify the output path for the PheWAS results file,
  phecode_file = "/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_phenotype_PheCode-matrix.txt",
  covariate_files = c("/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_covariates.txt", "/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.1/PMBB-Release-2020-2.1_phenotype_covariates.txt"),
  bfile = "/project/PMBB/PMBB-Release-2020-2.0/Exome/pVCF/all_variants/PMBB-Release-2020-2.0_genetic_exome_GL",
  plink_bin = "/project/voltron/Applications/PLINK/plink2_linux_avx2_20230607/plink2",
  cores = 16
)
#' }
```

  

## Render a report from pre-saved pheWAS output

    

  

```{r example-render_pmbb_phewas_report, eval = FALSE}
#' \dontrun{
render_pmbb_phewas_report()
#' }
```

  

  


```{r eval = FALSE, include = FALSE}
run_phewas_report <- function(gene, annotation_file, gene_col, masks, mask_operator, variant_id_col, effect_allele_col, plink_bin, bfile, phecode_file, covariate_files, populations, covariate_population_col, covariate_cols, mask_output_path = NULL, phewas_output_path = NULL, ...) {
  
  # Get the function call as a string
  function_call <- deparse(match.call())
  
  # return(function_call)
  
  # Extract genotypes for a series of masks
  mask_res <- pmbb_extract_genotype_masks(
    gene = gene,
    annotation_file = annotation_file,
    gene_col = gene_col,
    masks = masks,
    mask_operator = mask_operator,
    variant_id_col = {{ variant_id_col }},
    effect_allele_col = {{ effect_allele_col }},
    plink_bin = plink_bin,
    bfile = bfile
  )
  
  # Save the mask results to a file
  if (is.null(mask_output_path)) {
    # Use a temporary file if mask_output_path is not provided
    mask_output_path <- tempfile(fileext = ".rds")
  }
  saveRDS(mask_res, file = mask_output_path)
  
  # Run a PheWAS
  phewas_res <- run_pmbb_phewas(
    mask_genotypes_list = mask_res,
    phecode_file = phecode_file,
    covariate_files = covariate_files,
    populations = populations,
    covariate_population_col = {{ covariate_population_col }},
    covariate_cols = {{ covariate_cols }},
    ...
  )
  
  # Save the PheWAS results to a file
  if (is.null(phewas_output_path)) {
    # Use a temporary file if phewas_output_path is not provided
    phewas_output_path <- tempfile(fileext = ".rds")
  }
  saveRDS(phewas_res, file = phewas_output_path)
  
  # Collapse the gene names into a single string
  gene_names <- glue::glue_collapse(gene, last = ", and ", sep = ", ")
  
  # Render the Quarto document using the saved files, function call, and gene names as execute_params
  
  usethis::use_template(".phecode_phewas_template.qmd", package = "pmbbPheWASr")
  
  quarto::quarto_render(
    input = ".phecode_phewas_template.qmd",
    output_format = "html",
    execute_params = list(
      mask_results = fs::path_abs(mask_output_path),
      phewas_results = fs::path_abs(phewas_output_path),
      function_call = function_call,
      gene_names = gene_names
    ),
    output_file = paste0(gene_names, "_phewas_report.html")
  )
  
  # Return the paths to the saved mask and PheWAS results files
  return(list(
    mask_output_path = mask_output_path,
    phewas_output_path = phewas_output_path
  ))
}
```

```{r eval = FALSE, include = FALSE}
# Example usage
run_pmbb_phewas_report(
  gene = c("LDLR"),
  annotation_file = "/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt",
  gene_col = "Gene.refGene",
  masks = list(
    plof_0.001 = list(ExonicFunc.ensGene = "== 'stopgain'", gnomAD_exome_ALL = "< 0.001")
  ),
  mask_operator = list(
    plof_0.001 = "burden"
  ),
  variant_id_col = ID,
  effect_allele_col = Alt,
  plink_bin = "/project/voltron/Applications/PLINK/plink2_linux_avx2_20230607/plink2",
  bfile = "/project/PMBB/PMBB-Release-2020-2.0/Exome/pVCF/all_variants/PMBB-Release-2020-2.0_genetic_exome_GL",
  phecode_file = "/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_phenotype_PheCode-matrix.txt",
  covariate_files = c("/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_covariates.txt", "/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.1/PMBB-Release-2020-2.1_phenotype_covariates.txt"),
  populations = c("ALL", "EUR", "AFR"),
  covariate_population_col = "Class",
  covariate_cols = c(Age = Age_at_Enrollment, Sex = Gen_Sex, dplyr::starts_with("Genotype_PC")),
  mask_output_path = "LDLR_mask_results.rds",  # Optional: Specify the output path for the mask results file
  phewas_output_path = "LDLR_phewas_results.rds",  # Optional: Specify the output path for the PheWAS results file,
  cores = 16
)
```

```{r eval = FALSE, include = FALSE}
job::job({run_pmbb_phewas_report(
    gene = c("ANGPTL3", "ANGPTL4", "ANGPTL8", "APOC3"),
    annotation_file = "/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt",
    gene_col = "Gene.refGene",
    masks = list(
        plof_0.01 = list(ExonicFunc.ensGene = "%in% c('stopgain', 'stoploss', 'frameshift substitution')", gnomAD_exome_ALL = "< 0.01"),
        missense_0.01 = list(ExonicFunc.ensGene = "%in% c('nonsynonymous SNV')", gnomAD_exome_ALL = "< 0.01")
    ),
    mask_operator = list(
        plof_0.01 = "burden",
        missense_0.01 = "burden"
    ),
    variant_id_col = ID,
    effect_allele_col = Alt,
    populations = c("ALL", "EUR", "AFR"),
    covariate_population_col = "Class",
    covariate_cols = c(Age = Age_at_Enrollment, Sex = Gen_Sex, dplyr::starts_with("Genotype_PC")),
    mask_output_path = "TRL_mask_results.rds",  # Optional: Specify the output path for the mask results file
    phewas_output_path = "TRL_phewas_results.rds",  # Optional: Specify the output path for the PheWAS results file,
    phecode_file = "/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_phenotype_PheCode-matrix.txt",
    covariate_files = c("/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_covariates.txt", "/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.1/PMBB-Release-2020-2.1_phenotype_covariates.txt"),
    bfile = "/project/PMBB/PMBB-Release-2020-2.0/Exome/pVCF/all_variants/PMBB-Release-2020-2.0_genetic_exome_GL",
    plink_bin = "/project/voltron/Applications/PLINK/plink2_linux_avx2_20230607/plink2",
    cores = 16
)}, title = "TRL PheWAS")
```

```{r eval = FALSE, include = FALSE}
apoe_variants <- awk_str_filter(
  filename = "/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt",
  filter_col = "Gene.refGene",
  filter_str = c("APOE"))

apoe4_mask_res <- pmbb_extract_genotype_masks(
  gene = "APOE",
  annotation_file = "/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt",
  gene_col = "Gene.refGene",
  masks = list(
    apoe4_burden = list(avsnp150 = "%in% c('rs429358', 'rs7412')"),
    apoe_single = list(avsnp150 = "%in% c('rs429358', 'rs7412')")
  ),
  mask_operator = list(
    apoe4_burden = "burden",
    apoe_single = "single"
  ),
  variant_id_col = ID,
  effect_allele_col = Alt,
  plink_bin = "/project/voltron/Applications/PLINK/plink2_linux_avx2_20230607/plink2",
  bfile = "/project/PMBB/PMBB-Release-2020-2.0/Exome/pVCF/all_variants/PMBB-Release-2020-2.0_genetic_exome_GL"
)

job::job({run_pmbb_phewas_report(
    gene = c("APOE"),
    annotation_file = "/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt",
    gene_col = "Gene.refGene",
    masks = list(
      # apoe4_burden = list(avsnp150 = "%in% c('rs429358', 'rs7412')"),
      apoe_single = list(avsnp150 = "%in% c('rs429358', 'rs7412')")
    ),
    mask_operator = list(
      # apoe4_burden = "burden",
      apoe_single = "single"
    ),
    variant_id_col = ID,
    effect_allele_col = Alt,
    populations = c("ALL", "EUR", "AFR"),
    covariate_population_col = "Class",
    covariate_cols = c(Age = Age_at_Enrollment, Sex = Gen_Sex, dplyr::starts_with("Genotype_PC")),
    mask_output_path = "APOE_mask_results.rds",  # Optional: Specify the output path for the mask results file
    phewas_output_path = "APOE_phewas_results.rds",  # Optional: Specify the output path for the PheWAS results file,
    phecode_file = "/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_phenotype_PheCode-matrix.txt",
    covariate_files = c("/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_covariates.txt", "/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.1/PMBB-Release-2020-2.1_phenotype_covariates.txt"),
    bfile = "/project/PMBB/PMBB-Release-2020-2.0/Exome/pVCF/all_variants/PMBB-Release-2020-2.0_genetic_exome_GL",
    plink_bin = "/project/voltron/Applications/PLINK/plink2_linux_avx2_20230607/plink2",
    cores = 16
)}, title = "APOE PheWAS")
```

```{r eval = FALSE, include = FALSE}
  render_pmbb_phewas_report(
    phewas_output_path = "/project/damrauer_shared/Users/mglevin/pmbbPheWASr/APOE_phewas_results.rds",
    mask_output_path = "/project/damrauer_shared/Users/mglevin/pmbbPheWASr/APOE_mask_results.rds",
    output_file = "APOE_report.html"
  )

```

<!--

# There can be development actions

Create a chunk with 'development' actions

- The chunk needs to be named `development` or `dev`
- It contains functions that are used for package development only
- Note that you may want to store most of these functions in the 0-dev_history.Rmd file

These are only included in the present flat template file, their content will not be part of the package anywhere else.
-->


