<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Extract genotypes based on variant masks — pmbb_extract_genotype_masks • pmbbPheWASr</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Extract genotypes based on variant masks — pmbb_extract_genotype_masks"><meta property="og:description" content="This function extracts genotypes for a subset of participants based on a set of masks. It wraps awk_str_filter, apply_variant_masks, and pmbb_extract_genotypes to extract genotypes for PMBB participants based on a set of masks."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pmbbPheWASr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/get-started.html">Get started</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mglev1n/pmbbPheWASr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Extract genotypes based on variant masks</h1>
    <small class="dont-index">Source: <a href="https://github.com/mglev1n/pmbbPheWASr/blob/HEAD/R/pmbb_extract_genotype_masks.R" class="external-link"><code>R/pmbb_extract_genotype_masks.R</code></a></small>
    <div class="hidden name"><code>pmbb_extract_genotype_masks.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function extracts genotypes for a subset of participants based on a set of masks. It wraps <code>awk_str_filter</code>, <code>apply_variant_masks</code>, and <code>pmbb_extract_genotypes</code> to extract genotypes for PMBB participants based on a set of masks.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">pmbb_extract_genotype_masks</span><span class="op">(</span></span>
<span>  <span class="va">gene</span>,</span>
<span>  <span class="va">annotation_file</span>,</span>
<span>  <span class="va">gene_col</span>,</span>
<span>  <span class="va">masks</span>,</span>
<span>  <span class="va">variant_id_col</span>,</span>
<span>  <span class="va">effect_allele_col</span>,</span>
<span>  mask_operator <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">plink_bin</span>,</span>
<span>  <span class="va">bfile</span>,</span>
<span>  threads <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  memory <span class="op">=</span> <span class="fl">8000</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>gene</dt>
<dd><p>The name of the gene to filter variants for.</p></dd>


<dt>annotation_file</dt>
<dd><p>The path to the variant annotation file.</p></dd>


<dt>gene_col</dt>
<dd><p>The name of the column in the annotation file to filter the gene on.</p></dd>


<dt>masks</dt>
<dd><p>A named list of lists containing filter criteria for each mask.
Each element of the outer list should be named after the desired output name for the filtered data frame (eg. name of the variant mask).
Each inner list should contain the filter criteria for a specific set of columns.
For example: list(plof_0.001 = list(ExonicFunc.ensGene = "== 'stopgain'", gnomAD_exome_ALL = "&lt; 0.001"), common_0.01 = list(gnomAD_exome_ALL = "&gt; 0.01"))</p></dd>


<dt>variant_id_col</dt>
<dd><p>The name of the column in the annotation file that contains the variant IDs.</p></dd>


<dt>effect_allele_col</dt>
<dd><p>The name of the column in the annotation file that contains the effect alleles.</p></dd>


<dt>mask_operator</dt>
<dd><p>A named list specifying the operation to perform for each mask. Valid operations are "burden" (default) and "single".
The names in the mask_operator list should correspond to the names in the masks list.
For example: list(plof_0.001 = "burden", common_0.01 = "single")</p></dd>


<dt>plink_bin</dt>
<dd><p>The path to the <code>plink2</code> binary executable.</p></dd>


<dt>bfile</dt>
<dd><p>The prefix of the plink binary files (without the file extension).</p></dd>


<dt>threads</dt>
<dd><p>The number of threads to use for parallel processing (default: 1).</p></dd>


<dt>memory</dt>
<dd><p>The amount of memory (in MB) to allocate for <code>plink2</code> (default: 8000).</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A named list, where each element represents a mask and contains a list with components based on the specified mask_operator:
- If mask_operator is "burden" (default):
- "variants": A <code><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble::tibble()</a></code> of the variants included in the mask.
- "genotypes": A <code><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble::tibble()</a></code> containing the genotype for each PMBB participant.
- "mask_type": A <code><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble::tibble()</a></code> containing the mask type applied (eg. "burden" or "single")</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="va">ldlr_mask_res</span> <span class="op">&lt;-</span> <span class="fu">pmbb_extract_genotype_masks</span><span class="op">(</span></span></span>
<span class="r-in"><span>  gene <span class="op">=</span> <span class="st">"LDLR"</span>,</span></span>
<span class="r-in"><span>  annotation_file <span class="op">=</span> <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt"</span>,</span></span>
<span class="r-in"><span>  gene_col <span class="op">=</span> <span class="st">"Gene.refGene"</span>,</span></span>
<span class="r-in"><span>  masks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    plof_0.001 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ExonicFunc.ensGene <span class="op">=</span> <span class="st">"== 'stopgain'"</span>, gnomAD_exome_ALL <span class="op">=</span> <span class="st">"&lt; 0.001"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    rare_0.000001 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>gnomAD_exome_ALL <span class="op">=</span> <span class="st">"&lt; 0.000001"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">)</span>,</span></span>
<span class="r-in"><span>  mask_operator <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    plof_0.001 <span class="op">=</span> <span class="st">"burden"</span>,</span></span>
<span class="r-in"><span>    rare_0.000001 <span class="op">=</span> <span class="st">"single"</span></span></span>
<span class="r-in"><span>  <span class="op">)</span>,</span></span>
<span class="r-in"><span>  variant_id_col <span class="op">=</span> <span class="va">ID</span>,</span></span>
<span class="r-in"><span>  effect_allele_col <span class="op">=</span> <span class="va">Alt</span>,</span></span>
<span class="r-in"><span>  plink_bin <span class="op">=</span> <span class="st">"/project/voltron/Applications/PLINK/plink2_linux_avx2_20230607/plink2"</span>,</span></span>
<span class="r-in"><span>  bfile <span class="op">=</span> <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Exome/pVCF/all_variants/PMBB-Release-2020-2.0_genetic_exome_GL"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Michael Levin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

