<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function is a wrapper around the run_pmbb_phewas function that additionally generates an HTML report of the results."><title>Run a PheWAS and generate a report — run_pmbb_phewas_report • pmbbPheWASr</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Run a PheWAS and generate a report — run_pmbb_phewas_report"><meta property="og:description" content="This function is a wrapper around the run_pmbb_phewas function that additionally generates an HTML report of the results."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">pmbbPheWASr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/get-started.html">Get started</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mglev1n/pmbbPheWASr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Run a PheWAS and generate a report</h1>
      <small class="dont-index">Source: <a href="https://github.com/mglev1n/pmbbPheWASr/blob/HEAD/R/run_pmbb_phewas_report.R" class="external-link"><code>R/run_pmbb_phewas_report.R</code></a></small>
      <div class="d-none name"><code>run_pmbb_phewas_report.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function is a wrapper around the <code>run_pmbb_phewas</code> function that additionally generates an HTML report of the results.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">run_pmbb_phewas_report</span><span class="op">(</span></span>
<span>  <span class="va">gene</span>,</span>
<span>  <span class="va">annotation_file</span>,</span>
<span>  <span class="va">gene_col</span>,</span>
<span>  <span class="va">masks</span>,</span>
<span>  <span class="va">mask_operator</span>,</span>
<span>  <span class="va">variant_id_col</span>,</span>
<span>  <span class="va">effect_allele_col</span>,</span>
<span>  <span class="va">plink_bin</span>,</span>
<span>  <span class="va">bfile</span>,</span>
<span>  <span class="va">phecode_file</span>,</span>
<span>  <span class="va">covariate_files</span>,</span>
<span>  <span class="va">populations</span>,</span>
<span>  <span class="va">covariate_population_col</span>,</span>
<span>  <span class="va">covariate_cols</span>,</span>
<span>  mask_output_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  phewas_output_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  report_output_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>gene</dt>
<dd><p>The name of the gene to filter variants for.</p></dd>


<dt>annotation_file</dt>
<dd><p>The path to the variant annotation file.</p></dd>


<dt>gene_col</dt>
<dd><p>The name of the column in the annotation file to filter the gene on.</p></dd>


<dt>masks</dt>
<dd><p>A named list of lists containing filter criteria for each mask.
Each element of the outer list should be named after the desired output name for the filtered data frame (eg. name of the variant mask).
Each inner list should contain the filter criteria for a specific set of columns.
For example: list(plof_0.001 = list(ExonicFunc.ensGene = "== 'stopgain'", gnomAD_exome_ALL = "&lt; 0.001"), common_0.01 = list(gnomAD_exome_ALL = "&gt; 0.01"))</p></dd>


<dt>mask_operator</dt>
<dd><p>A named list specifying the operation to perform for each mask. Valid operations are "burden" (default) and "single".
The names in the mask_operator list should correspond to the names in the masks list.
For example: list(plof_0.001 = "burden", common_0.01 = "single")</p></dd>


<dt>variant_id_col</dt>
<dd><p>The name of the column in the annotation file that contains the variant IDs.</p></dd>


<dt>effect_allele_col</dt>
<dd><p>The name of the column in the annotation file that contains the effect alleles.</p></dd>


<dt>plink_bin</dt>
<dd><p>The path to the <code>plink2</code> binary executable.</p></dd>


<dt>bfile</dt>
<dd><p>The prefix of the plink binary files (without the file extension).</p></dd>


<dt>phecode_file</dt>
<dd><p>Path to PMBB phenotype file or a dataframe containing phenotype data used to run PheWAS</p></dd>


<dt>covariate_files</dt>
<dd><p>Path to PMBB covariate file(s) or a dataframe containing covariate data used to run PheWAS. If multiple files are provided, they will be joined by <code>PMBB_ID</code>, and values for duplicated columns will be preserved from the from the file that is specified earlier.</p></dd>


<dt>populations</dt>
<dd><p>A character vector of populations to run PheWAS on. Default is <code>c("ALL")</code></p></dd>


<dt>covariate_population_col</dt>
<dd><p>Column containing population labels in the covariate file or dataframe, required if <code>populations</code> is not <code>ALL</code></p></dd>


<dt>covariate_cols</dt>
<dd><p>Vector of columns in the covariate file or dataframe that should be used as covariates in the PheWAS</p></dd>


<dt>mask_output_path</dt>
<dd><p>Path to the saved genotype output for each mask</p></dd>


<dt>phewas_output_path</dt>
<dd><p>Path to the saved PheWAS output dataframe</p></dd>


<dt>report_output_path</dt>
<dd><p>(Optional) File name and path were the report should be saved. If not provided the report will be saved to the current directory</p></dd>


<dt>...</dt>
<dd><p>Additional arguments passed to <code><a href="https://rdrr.io/pkg/PheWAS/man/phewas.html" class="external-link">PheWAS::phewas()</a></code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A named list containing the paths to the output files</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="fu">run_pmbb_phewas_report</span><span class="op">(</span></span></span>
<span class="r-in"><span>  gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"LDLR"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  annotation_file <span class="op">=</span> <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt"</span>,</span></span>
<span class="r-in"><span>  gene_col <span class="op">=</span> <span class="st">"Gene.refGene"</span>,</span></span>
<span class="r-in"><span>  masks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    plof_0.01 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ExonicFunc.ensGene <span class="op">=</span> <span class="st">"%in% c('stopgain', 'stoploss', 'frameshift substitution')"</span>, gnomAD_exome_ALL <span class="op">=</span> <span class="st">"&lt; 0.01"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    missense_0.01 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ExonicFunc.ensGene <span class="op">=</span> <span class="st">"%in% c('nonsynonymous SNV')"</span>, gnomAD_exome_ALL <span class="op">=</span> <span class="st">"&lt; 0.01"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">)</span>,</span></span>
<span class="r-in"><span>  mask_operator <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    plof_0.01 <span class="op">=</span> <span class="st">"burden"</span>,</span></span>
<span class="r-in"><span>    missense_0.01 <span class="op">=</span> <span class="st">"burden"</span></span></span>
<span class="r-in"><span>  <span class="op">)</span>,</span></span>
<span class="r-in"><span>  variant_id_col <span class="op">=</span> <span class="va">ID</span>,</span></span>
<span class="r-in"><span>  effect_allele_col <span class="op">=</span> <span class="va">Alt</span>,</span></span>
<span class="r-in"><span>  populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ALL"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  covariate_population_col <span class="op">=</span> <span class="st">"Class"</span>,</span></span>
<span class="r-in"><span>  covariate_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Age <span class="op">=</span> <span class="va">Age_at_Enrollment</span>, Sex <span class="op">=</span> <span class="va">Gen_Sex</span>, <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">starts_with</a></span><span class="op">(</span><span class="st">"Genotype_PC"</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  mask_output_path <span class="op">=</span> <span class="st">"LDLR_mask_results.rds"</span>,  <span class="co"># Optional: Specify the output path for the mask results file</span></span></span>
<span class="r-in"><span>  phewas_output_path <span class="op">=</span> <span class="st">"LDLR_phewas_results.rds"</span>,  <span class="co"># Optional: Specify the output path for the PheWAS results file,</span></span></span>
<span class="r-in"><span>  phecode_file <span class="op">=</span> <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_phenotype_PheCode-matrix.txt"</span>,</span></span>
<span class="r-in"><span>  covariate_files <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_covariates.txt"</span>, <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.1/PMBB-Release-2020-2.1_phenotype_covariates.txt"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  bfile <span class="op">=</span> <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Exome/pVCF/all_variants/PMBB-Release-2020-2.0_genetic_exome_GL"</span>,</span></span>
<span class="r-in"><span>  plink_bin <span class="op">=</span> <span class="st">"/project/voltron/Applications/PLINK/plink2_linux_avx2_20230607/plink2"</span>,</span></span>
<span class="r-in"><span>  cores <span class="op">=</span> <span class="fl">16</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michael Levin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

