<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Get started â€¢ pmbbPheWASr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Get started">
<meta property="og:description" content="pmbbPheWASr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pmbbPheWASr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/get-started.html">Get started</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mglev1n/pmbbPheWASr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Get started</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mglev1n/pmbbPheWASr/blob/HEAD/vignettes/get-started.Rmd" class="external-link"><code>vignettes/get-started.Rmd</code></a></small>
      <div class="hidden name"><code>get-started.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mglev1n/pmbbPheWASr" class="external-link">pmbbPheWASr</a></span><span class="op">)</span></span></code></pre></div>
<!-- WARNING - This vignette is generated by {fusen} from dev/flat_pmbb_phewas.Rmd: do not edit by hand -->
<!-- Run this 'development' chunk -->
<!-- Store every call to library() that you need to explore your functions -->
<!--
 You need to run the 'description' chunk in the '0-dev_history.Rmd' file before continuing your code there.

If it is the first time you use {fusen}, after 'description', you can directly run the last chunk of the present file with inflate() inside.
-->
<!-- 
 Store your dataset in a directory named "inst/" at the root of your project.
 Use it for your tests in this Rmd thanks to `pkgload::load_all()` to make it available
and `system.file()` to read it in your examples.

- There already is a dataset in the "inst/" directory to be used in the examples below
-->
<div class="section level2">
<h2 id="filter-variant-annotations-file">Filter Variant Annotations File<a class="anchor" aria-label="anchor" href="#filter-variant-annotations-file"></a>
</h2>
<p>The <code>awk_str_filter</code> function can be used to filter
arbitrary rows from a tab-delimited text file using <code>awk</code>.
The function accepts the path to a tab-delimited text file, name of a
column to filter, and the string to filter for. The function returns a
tibble containing the filtered rows. This function can be used to filter
the PMBB variant annotations file to only include variants within the
specified gene(s).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span><span class="va">ldlr_variants</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/awk_str_filter.html">awk_str_filter</a></span><span class="op">(</span></span>
<span>  filename <span class="op">=</span> <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt"</span>,</span>
<span>  filter_col <span class="op">=</span> <span class="st">"Gene.refGene"</span>,</span>
<span>  filter_str <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"LDLR"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="apply-variant-masks-to-annotation-file">Apply variant masks to annotation file<a class="anchor" aria-label="anchor" href="#apply-variant-masks-to-annotation-file"></a>
</h2>
<p>The <code>apply_variant_masks</code> function applies variant mask
criteria to a data frame containing variant annotations. The function
can accept multiple sets of filter criteria, each defined as a list of
column names and filter conditions. The resulting filtered data frames
are returned as a named list.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span></span>
<span><span class="co"># Extract variants in a gene</span></span>
<span><span class="va">ldlr_variants</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/awk_str_filter.html">awk_str_filter</a></span><span class="op">(</span></span>
<span>  filename <span class="op">=</span> <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt"</span>,</span>
<span>  filter_col <span class="op">=</span> <span class="st">"Gene.refGene"</span>,</span>
<span>  filter_str <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"LDLR"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Apply the specified masks to the set of variants</span></span>
<span><span class="fu"><a href="../reference/apply_variant_masks.html">apply_variant_masks</a></span><span class="op">(</span></span>
<span>  <span class="va">ldlr_variants</span>,</span>
<span>  masks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    plof_0.001 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ExonicFunc.ensGene <span class="op">=</span> <span class="st">"== 'stopgain'"</span>, gnomAD_exome_ALL <span class="op">=</span> <span class="st">"&lt; 0.001"</span><span class="op">)</span>,</span>
<span>    common_0.01 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>gnomAD_exome_ALL <span class="op">=</span> <span class="st">"&gt; 0.01"</span><span class="op">)</span>,</span>
<span>    exonic_0.001 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Func.refGene <span class="op">=</span> <span class="st">"== 'exonic'"</span>, gnomAD_exome_ALL <span class="op">=</span> <span class="st">"&lt; 0.001"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="extracting-genotypes">Extracting genotypes<a class="anchor" aria-label="anchor" href="#extracting-genotypes"></a>
</h2>
<div class="section level3">
<h3 id="extract-a-single-set-of-variants">Extract a single set of variants<a class="anchor" aria-label="anchor" href="#extract-a-single-set-of-variants"></a>
</h3>
<p>The <code>pmbb_extract_genotypes</code> function is a wrapper around
<code>plink2</code> to extract genotypes from PMBB plink files. The
function accepts the path to the plink files and the list of variant IDs
to extract. The function returns a long-format tibble containing the
extracted genotypes for each PMBB participant.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span><span class="fu"><a href="../reference/pmbb_extract_genotypes.html">pmbb_extract_genotypes</a></span><span class="op">(</span></span>
<span>  variant_df <span class="op">=</span> <span class="va">ldlr_variants</span>,</span>
<span>  variant_id_col <span class="op">=</span> <span class="va">ID</span>,</span>
<span>  effect_allele_col <span class="op">=</span> <span class="va">Alt</span>,</span>
<span>  plink_bin <span class="op">=</span> <span class="st">"/project/voltron/Applications/PLINK/plink2_linux_avx2_20230607/plink2"</span>,</span>
<span>  bfile <span class="op">=</span> <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Exome/pVCF/all_variants/PMBB-Release-2020-2.0_genetic_exome_GL"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="extract-genotypes-based-on-masks">Extract genotypes based on masks<a class="anchor" aria-label="anchor" href="#extract-genotypes-based-on-masks"></a>
</h3>
<p>The <code>pmbb_extract_genotype_masks</code> function extracts
genotypes for a subset of participants based on a set of masks. This
function wraps <code>awk_str_filter</code>,
<code>apply_variant_masks</code>, and
<code>pmbb_extract_genotypes</code> to extract genotypes for PMBB
participants based on a set of masks.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span><span class="va">ldlr_mask_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pmbb_extract_genotype_masks.html">pmbb_extract_genotype_masks</a></span><span class="op">(</span></span>
<span>  gene <span class="op">=</span> <span class="st">"LDLR"</span>,</span>
<span>  annotation_file <span class="op">=</span> <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt"</span>,</span>
<span>  gene_col <span class="op">=</span> <span class="st">"Gene.refGene"</span>,</span>
<span>  masks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    plof_0.001 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ExonicFunc.ensGene <span class="op">=</span> <span class="st">"== 'stopgain'"</span>, gnomAD_exome_ALL <span class="op">=</span> <span class="st">"&lt; 0.001"</span><span class="op">)</span>,</span>
<span>    rare_0.000001 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>gnomAD_exome_ALL <span class="op">=</span> <span class="st">"&lt; 0.000001"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  mask_operator <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    plof_0.001 <span class="op">=</span> <span class="st">"burden"</span>,</span>
<span>    rare_0.000001 <span class="op">=</span> <span class="st">"single"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  variant_id_col <span class="op">=</span> <span class="va">ID</span>,</span>
<span>  effect_allele_col <span class="op">=</span> <span class="va">Alt</span>,</span>
<span>  plink_bin <span class="op">=</span> <span class="st">"/project/voltron/Applications/PLINK/plink2_linux_avx2_20230607/plink2"</span>,</span>
<span>  bfile <span class="op">=</span> <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Exome/pVCF/all_variants/PMBB-Release-2020-2.0_genetic_exome_GL"</span><span class="op">)</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="phewas-convenience-functions">PheWAS convenience functions<a class="anchor" aria-label="anchor" href="#phewas-convenience-functions"></a>
</h2>
<div class="section level3">
<h3 id="format-pmbb-phecode-data">Format PMBB phecode data<a class="anchor" aria-label="anchor" href="#format-pmbb-phecode-data"></a>
</h3>
<p>The <code>pmbb_format_phecodes</code> function takes a path to a PMBB
phecode file and applies a case threshold to categorize individuals as
cases or controls. The function returns a tibble with
<code>PMBB_ID</code> and phecodes as columns, with TRUE/FALSE values for
each phecode column.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span><span class="fu"><a href="../reference/pmbb_format_phecodes.html">pmbb_format_phecodes</a></span><span class="op">(</span>phecode_file <span class="op">=</span> <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_phenotype_PheCode-matrix.txt"</span><span class="op">)</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-a-phewas">Run a PheWAS<a class="anchor" aria-label="anchor" href="#run-a-phewas"></a>
</h3>
<p>The <code>run_pmbb_phewas</code> function runs a PheWAS on the PMBB
data, using the results of the <code>pmbb_extract_genotype_masks</code>
function and paths to PMBB files containing phenotype and covariate
data.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span><span class="co"># Extract genotypes for a series of masks</span></span>
<span><span class="va">ldlr_mask_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pmbb_extract_genotype_masks.html">pmbb_extract_genotype_masks</a></span><span class="op">(</span></span>
<span>  gene <span class="op">=</span> <span class="st">"LDLR"</span>,</span>
<span>  annotation_file <span class="op">=</span> <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt"</span>,</span>
<span>  gene_col <span class="op">=</span> <span class="st">"Gene.refGene"</span>,</span>
<span>  masks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    plof_0.001 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ExonicFunc.ensGene <span class="op">=</span> <span class="st">"== 'stopgain'"</span>, gnomAD_exome_ALL <span class="op">=</span> <span class="st">"&lt; 0.001"</span><span class="op">)</span>,</span>
<span>    common <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>gnomAD_exome_ALL <span class="op">=</span> <span class="st">"&gt; 0.01"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  mask_operator <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    plof_0.001 <span class="op">=</span> <span class="st">"burden"</span>,</span>
<span>    common <span class="op">=</span> <span class="st">"single"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  variant_id_col <span class="op">=</span> <span class="va">ID</span>,</span>
<span>  effect_allele_col <span class="op">=</span> <span class="va">Alt</span>,</span>
<span>  plink_bin <span class="op">=</span> <span class="st">"/project/voltron/Applications/PLINK/plink2_linux_avx2_20230607/plink2"</span>,</span>
<span>  bfile <span class="op">=</span> <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Exome/pVCF/all_variants/PMBB-Release-2020-2.0_genetic_exome_GL"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run a PheWAS</span></span>
<span><span class="va">phewas_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pmbb_phewas.html">run_pmbb_phewas</a></span><span class="op">(</span></span>
<span>  mask_genotypes_list <span class="op">=</span> <span class="va">ldlr_mask_res</span>,</span>
<span>  phecode_file <span class="op">=</span> <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_phenotype_PheCode-matrix.txt"</span>,</span>
<span>  covariate_files <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_covariates.txt"</span>, <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.1/PMBB-Release-2020-2.1_phenotype_covariates.txt"</span><span class="op">)</span>,</span>
<span>  populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ALL"</span>, <span class="st">"EUR"</span><span class="op">)</span>,</span>
<span>  covariate_population_col <span class="op">=</span> <span class="va">Class</span>,</span>
<span>  covariate_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Age <span class="op">=</span> <span class="va">Age_at_Enrollment</span>, Sex <span class="op">=</span> <span class="va">Gen_Sex</span>, <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">starts_with</a></span><span class="op">(</span><span class="st">"Genotype_PC"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-a-phewas-and-generate-a-report">Run a PheWAS and generate a report<a class="anchor" aria-label="anchor" href="#run-a-phewas-and-generate-a-report"></a>
</h3>
<p>The <code>run_pmbb_phewas_report</code> function is a wrapper around
the <code>run_pmbb_phewas</code> function that generates an HTML report
of the results.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span><span class="fu"><a href="../reference/run_pmbb_phewas_report.html">run_pmbb_phewas_report</a></span><span class="op">(</span></span>
<span>  gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"LDLR"</span><span class="op">)</span>,</span>
<span>  annotation_file <span class="op">=</span> <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt"</span>,</span>
<span>  gene_col <span class="op">=</span> <span class="st">"Gene.refGene"</span>,</span>
<span>  masks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    plof_0.01 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ExonicFunc.ensGene <span class="op">=</span> <span class="st">"%in% c('stopgain', 'stoploss', 'frameshift substitution')"</span>, gnomAD_exome_ALL <span class="op">=</span> <span class="st">"&lt; 0.01"</span><span class="op">)</span>,</span>
<span>    missense_0.01 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ExonicFunc.ensGene <span class="op">=</span> <span class="st">"%in% c('nonsynonymous SNV')"</span>, gnomAD_exome_ALL <span class="op">=</span> <span class="st">"&lt; 0.01"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  mask_operator <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    plof_0.01 <span class="op">=</span> <span class="st">"burden"</span>,</span>
<span>    missense_0.01 <span class="op">=</span> <span class="st">"burden"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  variant_id_col <span class="op">=</span> <span class="va">ID</span>,</span>
<span>  effect_allele_col <span class="op">=</span> <span class="va">Alt</span>,</span>
<span>  populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ALL"</span><span class="op">)</span>,</span>
<span>  covariate_population_col <span class="op">=</span> <span class="st">"Class"</span>,</span>
<span>  covariate_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Age <span class="op">=</span> <span class="va">Age_at_Enrollment</span>, Sex <span class="op">=</span> <span class="va">Gen_Sex</span>, <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">starts_with</a></span><span class="op">(</span><span class="st">"Genotype_PC"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  mask_output_path <span class="op">=</span> <span class="st">"LDLR_mask_results.rds"</span>,  <span class="co"># Optional: Specify the output path for the mask results file</span></span>
<span>  phewas_output_path <span class="op">=</span> <span class="st">"LDLR_phewas_results.rds"</span>,  <span class="co"># Optional: Specify the output path for the PheWAS results file,</span></span>
<span>  phecode_file <span class="op">=</span> <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_phenotype_PheCode-matrix.txt"</span>,</span>
<span>  covariate_files <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_covariates.txt"</span>, <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.1/PMBB-Release-2020-2.1_phenotype_covariates.txt"</span><span class="op">)</span>,</span>
<span>  bfile <span class="op">=</span> <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Exome/pVCF/all_variants/PMBB-Release-2020-2.0_genetic_exome_GL"</span>,</span>
<span>  plink_bin <span class="op">=</span> <span class="st">"/project/voltron/Applications/PLINK/plink2_linux_avx2_20230607/plink2"</span>,</span>
<span>  cores <span class="op">=</span> <span class="fl">16</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="render-a-report-from-pre-saved-phewas-output">Render a report from pre-saved pheWAS output<a class="anchor" aria-label="anchor" href="#render-a-report-from-pre-saved-phewas-output"></a>
</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' \dontrun{</span></span>
<span><span class="fu"><a href="../reference/render_pmbb_phewas_report.html">render_pmbb_phewas_report</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#' }</span></span></code></pre></div>
<!--

# There can be development actions

Create a chunk with 'development' actions

- The chunk needs to be named `development` or `dev`
- It contains functions that are used for package development only
- Note that you may want to store most of these functions in the 0-dev_history.Rmd file

These are only included in the present flat template file, their content will not be part of the package anywhere else.
-->
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Michael Levin.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
