<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="pmbbPheWASr">
<title>Get started • pmbbPheWASr</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get started">
<meta property="og:description" content="pmbbPheWASr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">pmbbPheWASr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.9</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/pmbbPheWASr.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mglev1n/pmbbPheWASr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Get started</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mglev1n/pmbbPheWASr/blob/HEAD/vignettes/pmbbPheWASr.Rmd" class="external-link"><code>vignettes/pmbbPheWASr.Rmd</code></a></small>
      <div class="d-none name"><code>pmbbPheWASr.Rmd</code></div>
    </div>

    
    
<p><a href="https://doi.org/10.1093%2Fbioinformatics%2Fbtq126" class="external-link">Phenome-wide
association studies (PheWAS)</a>) are an approach to systematically
investigate the association between genetic variant(s) and a wide range
of phenotypes. The <code>pmbbPheWASr</code> package is a collection of
functions to perform PheWAS analyses using data from the Penn Medicine
Biobank. The package is designed to be user-friendly and flexible,
allowing users to perform PheWAS analyses with minimal effort. The
function contains several convenience functions that facilitate running
PheWAS using PMBB data, including extracting genotype information based
on user-supplied masks, formatting phenotype/covariate data, and running
PheWAS analyses using the <a href="https://github.com/PheWAS/PheWAS" class="external-link"><code>PheWAS</code>
package</a>.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"mglev1n/pmbbPheWASr"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mglev1n/pmbbPheWASr" class="external-link">pmbbPheWASr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="extract-genotypes-for-a-specified-genotype-mask">Extract genotypes for a specified genotype mask<a class="anchor" aria-label="anchor" href="#extract-genotypes-for-a-specified-genotype-mask"></a>
</h2>
<p>When performing PheWAS with whole-genome or whole-exome sequence
data, the frequency of some individual variants within the population
may be very low, limiting our ability to test for associations of these
variants with phenotypes. To address this issue, one approach is to
consider the “burden” of genetic variation within a region/gene/pathway
that fit certain criteria. For example, we may want to consider the
burden of rare loss-of-function variants with a minor allele frequency
(MAF) less than 1% within a gene. The
<code>pmbb_extract_genotype_masks</code> function allows users to
extract genotypes for one or more specified genotype masks.</p>
<p>Here, we will use the <code>masks</code> argument to define a mask
called <code>plof_lt_0.001</code> that includes all putative
loss-of-function variants (‘stopgain’, ‘stoploss’, ‘frameshift
substitution’) with a gnoMAD minor allele frequency less than 0.1% in
the <code>CFTR</code> gene. We will then use the
<code>mask_operator</code> argument to specify that we want to consider
the <code>burden</code> of variants in this mask.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cftr_mask_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pmbb_extract_genotype_masks.html">pmbb_extract_genotype_masks</a></span><span class="op">(</span></span>
<span>  gene <span class="op">=</span> <span class="st">"CFTR"</span>,</span>
<span>  annotation_file <span class="op">=</span> <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt"</span>,</span>
<span>  gene_col <span class="op">=</span> <span class="st">"Gene.refGene"</span>,</span>
<span>  masks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    plof_lt_0.001 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ExonicFunc.ensGene <span class="op">=</span> <span class="st">"%in% c('stopgain', 'stoploss', 'frameshift substitution')"</span>, gnomAD_exome_ALL <span class="op">=</span> <span class="st">"&lt; 0.001"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  mask_operator <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    plof_lt_0.001 <span class="op">=</span> <span class="st">"burden"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  variant_id_col <span class="op">=</span> <span class="va">ID</span>,</span>
<span>  effect_allele_col <span class="op">=</span> <span class="va">Alt</span>,</span>
<span>  plink_bin <span class="op">=</span> <span class="st">"/project/voltron/Applications/PLINK/plink2_linux_avx2_20230607/plink2"</span>,</span>
<span>  bfile <span class="op">=</span> <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Exome/pVCF/all_variants/PMBB-Release-2020-2.0_genetic_exome_GL"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We might also want to individually consider any Pathogenic/Likely
pathogenic variants that are present in &gt;10 PMBB participants. We can
define a new mask called <code>p_lp_het_gt_10</code> that includes all
Pathogenic/Likely pathogenic variants in the <code>CFTR</code> gene that
are present in &gt;10 PMBB participants. We will then use the
<code>mask_operator</code> argument to specify that we want to
individually consider each <code>single</code> variant in this mask.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cftr_mask_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pmbb_extract_genotype_masks.html">pmbb_extract_genotype_masks</a></span><span class="op">(</span></span>
<span>  gene <span class="op">=</span> <span class="st">"CFTR"</span>,</span>
<span>  annotation_file <span class="op">=</span> <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt"</span>,</span>
<span>  gene_col <span class="op">=</span> <span class="st">"Gene.refGene"</span>,</span>
<span>  masks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    plof_lt_0.001 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ExonicFunc.ensGene <span class="op">=</span> <span class="st">"%in% c('stopgain', 'stoploss', 'frameshift substitution')"</span>, gnomAD_exome_ALL <span class="op">=</span> <span class="st">"&lt; 0.001"</span><span class="op">)</span>,</span>
<span>    p_lp_het_gt_10 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>CLNSIG <span class="op">=</span> <span class="st">"%in% c('Pathogenic', 'Pathogenic/Likely_pathogenic')"</span>, HET_REF_ALT_CTS <span class="op">=</span> <span class="st">"&gt; 10"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  mask_operator <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    plof_lt_0.001 <span class="op">=</span> <span class="st">"burden"</span>,</span>
<span>    p_lp_het_gt_10 <span class="op">=</span> <span class="st">"single"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  variant_id_col <span class="op">=</span> <span class="va">ID</span>,</span>
<span>  effect_allele_col <span class="op">=</span> <span class="va">Alt</span>,</span>
<span>  plink_bin <span class="op">=</span> <span class="st">"/project/voltron/Applications/PLINK/plink2_linux_avx2_20230607/plink2"</span>,</span>
<span>  bfile <span class="op">=</span> <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Exome/pVCF/all_variants/PMBB-Release-2020-2.0_genetic_exome_GL"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The <code>pmbb_extract_genotype_masks</code> function is designed to
be flexible. The examples above use the default variant annotation file
from PMBB, but users can also provide their own variant annotation
file.</p>
<p>The output of the <code>pmbb_extract_genotype_masks</code> function
is a list, where each element corresponds to a mask or mask-variant
combination (in the case when individual variants are to be tested).
Each element of the list contains information about the mask criteria
that were applied, the variants that were extracted, and the
corresponding genotypes (either individually or as a burden, depending
on the <code>mask_operator</code> arguments).</p>
</div>
<div class="section level2">
<h2 id="format-phenotypecovariate-data">Format phenotype/covariate data<a class="anchor" aria-label="anchor" href="#format-phenotypecovariate-data"></a>
</h2>
<p>The following functions allow users to format phenotype and covariate
data for use in PheWAS analyses.</p>
<div class="section level3">
<h3 id="format-phecodes">Format PheCodes<a class="anchor" aria-label="anchor" href="#format-phecodes"></a>
</h3>
<p>The <code>pmbb_format_phecodes</code> function allows users to format
the PMBB PheCode data for use in PheWAS analyses. The
<code>phecode_case_threshold</code> allows the user to define how many
instances of a phecode are required for an individual to be considered a
case. For example, if <code>phecode_case_threshold = 2</code>, then a
participant must have at least 2 instances of a phecode to be considered
a case for that phecode:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pmbb_phecodes_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pmbb_format_phecodes.html">pmbb_format_phecodes</a></span><span class="op">(</span>phecode_file <span class="op">=</span> <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_phenotype_PheCode-matrix.txt"</span>, phecode_case_threshold <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="format-labs">Format labs<a class="anchor" aria-label="anchor" href="#format-labs"></a>
</h3>
<p>The <code>pmbb_format_labs</code> function allows users to format
PMBB lab data for use in PheWAS analyses. The user can specify which lab
files to extract from the lab data, and specify whether inpatient or
outpatient labs should be used:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pmbb_labs_df</span> <span class="op">&lt;-</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_ls.html" class="external-link">dir_ls</a></span><span class="op">(</span><span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/"</span>, glob <span class="op">=</span> <span class="st">"*labs*"</span>, recurse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/pmbb_format_labs.html">pmbb_format_labs</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="format-covariates">Format covariates<a class="anchor" aria-label="anchor" href="#format-covariates"></a>
</h3>
<p>The <code>pmbb_format_covariates</code> function allows users to
format PMBB demographic and covariate data. The user may supply multiple
files containing covariate data, and they will be sequentially appended
to the first file. Users can also specify which covariates should be
extracted from the files, and any population subgroups that should be
included in stratified analysis.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pmbb_covariates_df</span> <span class="op">&lt;-</span> <span class="fu">pmbbPheWASr</span><span class="fu">:::</span><span class="fu">pmbb_format_covariates</span><span class="op">(</span></span>
<span>  covariate_files <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_covariates.txt"</span>, <span class="st">"/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.1/PMBB-Release-2020-2.1_phenotype_covariates.txt"</span><span class="op">)</span>,</span>
<span>  populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ALL"</span>, <span class="st">"EUR"</span><span class="op">)</span>,</span>
<span>  covariate_population_col <span class="op">=</span> <span class="va">Class</span>,</span>
<span>  covariate_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Age <span class="op">=</span> <span class="va">Age_at_Enrollment</span>, Sex <span class="op">=</span> <span class="va">Gen_Sex</span>, <span class="fu">starts_with</span><span class="op">(</span><span class="st">"Genotype_PC"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="run-a-phewas">Run a PheWAS<a class="anchor" aria-label="anchor" href="#run-a-phewas"></a>
</h2>
<p>The <code>run_pmbb_phewas</code> function runs a PheWAS on the PMBB
data. The user may specify the genotype, phenotype, and covariate data
to include in the analysis. Here, we’ll first combine the phecode and
lab data:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pmbb_phenotypes_combined</span> <span class="op">&lt;-</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">pmbb_phecodes_df</span>, <span class="va">pmbb_labs_df</span>, by <span class="op">=</span> <span class="st">"PMBB_ID"</span><span class="op">)</span></span></code></pre></div>
<p>Then, we’ll perform a PheWAS of the different <code>CFTR</code>
genotype masks across both the phecode and lab datasets. Depending on
the analysis, this may be very computationally demanding. The PheWAS can
be spread over several cores by specifying the <code>cores</code>
argument. This is useful when running analyses in the Penn HPC/LPC
environments, where multiple cores can be requested. For testing
purposes, we’ll limit ourselves to the cystic fibrosis phenotype
(phecode = “499”):</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cftr_phewas_res</span> <span class="op">&lt;-</span> <span class="fu">pmbbPheWASr</span><span class="fu">::</span><span class="fu"><a href="../reference/run_pmbb_phewas.html">run_pmbb_phewas</a></span><span class="op">(</span></span>
<span>  mask_genotypes_list <span class="op">=</span> <span class="va">cftr_mask_res</span>,</span>
<span>  populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ALL"</span>, <span class="st">"EUR"</span>, <span class="st">"AFR"</span><span class="op">)</span>,</span>
<span>  covariate_population_col <span class="op">=</span> <span class="st">"Class"</span>,</span>
<span>  covariate_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Age <span class="op">=</span> <span class="va">Age_at_Enrollment</span>, Sex <span class="op">=</span> <span class="va">Gen_Sex</span>, <span class="fu">starts_with</span><span class="op">(</span><span class="st">"Genotype_PC"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  phenotypes <span class="op">=</span> <span class="va">pmbb_phenotypes_combined</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">select</span><span class="op">(</span><span class="va">PMBB_ID</span>, <span class="st">"499"</span><span class="op">)</span>,</span>
<span>  covariates <span class="op">=</span> <span class="va">pmbb_covariates_df</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualize-phewas-results">Visualize PheWAS results<a class="anchor" aria-label="anchor" href="#visualize-phewas-results"></a>
</h2>
<p>The <code>render_pmbb_phewas_report</code> function can be used to
generate a report containing the PheWAS results:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/render_pmbb_phewas_report.html">render_pmbb_phewas_report</a></span><span class="op">(</span></span>
<span>  phewas_output <span class="op">=</span> <span class="va">cftr_phewas_res</span>,</span>
<span>  mask_output <span class="op">=</span> <span class="va">cftr_mask_res</span>,</span>
<span>  output_file <span class="op">=</span> <span class="st">"CFTR_pheWAS_report.html"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The report will use the internal template by default, but you can
also specify a template using the <code>template_path</code> argument.
You can generate a copy of the default template using
<code>usethis::use_template("combined_phewas_template.qmd", save_as = "combined_phewas_template.qmd", package = "pmbbPheWASr")</code>,
which will be saved in the current working directory. You can then edit
this template as needed and specify it in the <code>template_path</code>
argument.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Michael Levin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
