# WARNING - Generated by {fusen} from dev/flat_pmbb_phewas.Rmd: do not edit by hand

#' Filter a Data Frame Based on Multiple Criteria
#' 
#' This function filters a data frame based on a list of column names and corresponding filter criteria.
#' The filter criteria are provided as a list, where each element represents a column and its associated filter condition.
#' The function applies the filter criteria to the specified columns and returns the filtered data frame.
#' 
#' @param df A data frame to be filtered.
#' @param filter_list A list containing column names and their corresponding filter criteria.
#'   Each element of the list should be named after the column to be filtered and contain the filter condition as a string.
#'   For example: list(age = "> 30", city = "== 'New York'")
#'   
#' @return The filtered [tibble::tibble()] based on the provided filter criteria.
#' 
#' @import dplyr
#' @import rlang
#' @noRd
#' @examples
#' filter_dataframe(dplyr::starwars, list(height = " > 100", eye_color = " == 'blue'"))

filter_dataframe <- function(df, filter_list) {
  # Check if the input is a data frame
  if (!is.data.frame(df)) {
    abort("Input must be a data frame.")
  }
  
  # Check if the filter list is a list
  if (!is.list(filter_list)) {
    abort("Filter criteria must be provided as a list.")
  }
  
  # Iterate over each element in the filter list
  for (col_filter in names(filter_list)) {
    # Extract the filter criteria
    filter_criteria <- filter_list[[col_filter]]
    
    # Check if the column exists in the data frame
    if (!(col_filter %in% colnames(df))) {
      warn(glue::glue("Column '{col_filter}' not found in the data frame. Skipping filter for this column."))
      next
    }
    
    # Apply the filter criteria to the specified column using {{}}
    df <- df %>%
      filter(!!parse_expr(paste0({{col_filter}}, filter_criteria)))
  }
  
  return(df)
}

#' Extract genotypes from PMBB plink files
#' 
#' This function is a wrapper around `plink2` to extract genotypes from PMBB plink files. The function accepts the path to the plink files and the list of variant IDs to extract. The function returns a long-format [tibble::tibble()] containing the extracted genotypes for each PMBB participant.
#' 
#' @param variant_df A data frame containing the variant information.
#' @param variant_id_col The name of the column in `variant_df` that contains the variant IDs.
#' @param effect_allele_col The name of the column in `variant_df` that contains the effect alleles.
#' @param plink_bin The path to the `plink2` binary executable.
#' @param bfile The prefix of the plink binary files (without the file extension).
#' @param threads The number of threads to use for parallel processing (default: 1).
#' @param memory The amount of memory (in MB) to allocate for `plink2` (default: 8000).
#' 
#' @return A long-format [tibble::tibble()] containing the extracted genotypes for each PMBB participant.
#' 
#' @export
#' @examples
#' \dontrun{
#' pmbb_extract_genotypes(variant_df = ldlr_variants,
#'                        variant_id_col = ID,
#'                        effect_allele_col = Alt,
#'                        plink_bin = "/project/voltron/Applications/PLINK/plink2_linux_avx2_20230607/plink2",
#'                        bfile = "/project/PMBB/PMBB-Release-2020-2.0/Exome/pVCF/all_variants/PMBB-Release-2020-2.0_genetic_exome_GL")
#' }
pmbb_extract_genotypes <- function(variant_df, variant_id_col, effect_allele_col, plink_bin, bfile, threads = 1, memory = 8000) {
  
    cli::cli_progress_step("Extracting genotypes from {.file {bfile}}")
  
    allele_file <- fs::file_temp()
    genotype_file <- fs::path(fs::path_temp(), "genotypes")
    
    allele_table <- variant_df %>%
        dplyr::select({{variant_id_col}}, {{effect_allele_col}}) 
    
   allele_table %>%
        data.table::fwrite(allele_file, col.names = FALSE, sep = "\t")
    
    plink_result <- processx::run(
        command = plink_bin,
        args = c(
            "--bfile", bfile,
            "--threads", threads,
            "--memory", memory,
            "--snps", variant_df %>% dplyr::pull({{variant_id_col}}),
            "--export-allele", allele_file,
            "--export", "A",
            "--out", genotype_file
        ),
        error_on_status = TRUE
    )
    
    if (plink_result$status != 0) {
        cli::cli_abort("Error executing plink command. Status: {plink_result$stderr}")
    }
    
    genotype_raw_file <- paste0(genotype_file, ".raw")
    
    if (!file.exists(genotype_raw_file)) {
        cli::cli_abort("Genotype file not found: {genotype_raw_file}")
    }
    
    genotype_df <- data.table::fread(genotype_raw_file)
    
    genotype_tbl <- genotype_df %>%
        tibble::as_tibble() %>%
        tidytable::pivot_longer(cols = -c(`FID`, `IID`, `PAT`, `MAT`, `SEX`, `PHENOTYPE`),
                            names_to = "variant_id",
                            values_to = "genotype") %>%
        dplyr::select(PMBB_ID = IID, variant_id, genotype) %>%
      dplyr::as_tibble()
    
    return(genotype_tbl)
}
