# WARNING - Generated by {fusen} from dev/flat_pmbb_phewas.Rmd: do not edit by hand

#' Run a PheWAS on PMBB data
#' 
#' This function runs a PheWAS on PMBB data, using genotype information and paths to PMBB files containing phenotype and covariate data. The function ideally accepts genotype information from the `pmbb_extract_genotype_masks` function, but can accept any genotype information of the format `list(mask_name = list(genotypes = data.frame(PMBB_ID = character(), genotype = character())))`
#' 
#' @param mask_genotypes_list A named list, where each element contains a named element called `genotypes` which contains `PMBB_ID` and `genotype` columns.
#' @param phecode_file Path to PMBB phenotype file or a dataframe containing phenotype data used to run PheWAS
#' @param covariate_files Path to PMBB covariate file(s) or a dataframe containing covariate data used to run PheWAS. If multiple files are provided, they will be joined by `PMBB_ID`, and values for duplicated columns will be preserved from the from the file that is specified earlier.
#' @param populations A character vector of populations to run PheWAS on. Default is `c("ALL")`
#' @param covariate_cols Vector of columns in the covariate file or dataframe that should be used as covariates in the PheWAS
#' @param covariate_population_col Column containing population labels in the covariate file or dataframe, required if `populations` is not `ALL`
#' @param phenotype_case_count The minimum number of codes required to be considered a case (Default: 2)
#' @param ... Additional arguments passed to [PheWAS::phewas()]

#' @return A [tibble::tibble()] containing the results of the PheWAS for each population and mask
#' 
#' @export
#' @examples
#' \dontrun{
#' # Extract genotypes for a series of masks
#' ldlr_mask_res <- pmbb_extract_genotype_masks(
#'   gene = "LDLR",
#'   annotation_file = "/project/PMBB/PMBB-Release-2020-2.0/Exome/Variant_annotations/PMBB-Release-2020-2.0_genetic_exome_variant-annotation-counts.txt",
#'   gene_col = "Gene.refGene",
#'   masks = list(
#'     plof_0.001 = list(ExonicFunc.ensGene = "== 'stopgain'", gnomAD_exome_ALL = "< 0.001"),
#'     common = list(gnomAD_exome_ALL = "> 0.01")
#'   ),
#'   mask_operator = list(
#'     plof_0.001 = "burden",
#'     common = "single"
#'   ),
#'   variant_id_col = ID,
#'   effect_allele_col = Alt,
#'   plink_bin = "/project/voltron/Applications/PLINK/plink2_linux_avx2_20230607/plink2",
#'   bfile = "/project/PMBB/PMBB-Release-2020-2.0/Exome/pVCF/all_variants/PMBB-Release-2020-2.0_genetic_exome_GL")
#'
#' # Run a PheWAS
#' phewas_res <- run_pmbb_phewas(
#'   mask_genotypes_list = ldlr_mask_res,
#'   phecode_file = "/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_phenotype_PheCode-matrix.txt",
#'   covariate_files = c("/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.3/PMBB-Release-2020-2.3_covariates.txt", "/project/PMBB/PMBB-Release-2020-2.0/Phenotype/2.1/PMBB-Release-2020-2.1_phenotype_covariates.txt"),
#'   populations = c("ALL", "EUR"),
#'   covariate_population_col = Class,
#'   covariate_cols = c(Age = Age_at_Enrollment, Sex = Gen_Sex, dplyr::starts_with("Genotype_PC"))
#' )
#' }
run_pmbb_phewas <- function(mask_genotypes_list, phecode_file, covariate_files, populations = c("ALL"), covariate_cols, covariate_population_col = NULL, phenotype_case_count = 2, ...) {
  # Check if mask_genotypes_list is a list
  if (!is.list(mask_genotypes_list)) {
    cli::cli_abort("{.arg mask_genotypes_list} must be a list")
  }

  # Check if phenotype_file is a file path or a dataframe
  if (is.character(phecode_file)) {
    if (!file.exists(phecode_file)) {
      cli::cli_abort("{phenotype_file} does not exist")
    }
    phecode_df <- pmbb_format_phecodes(phecode_file, phenotype_case_count)
  } else if (is.data.frame(phecode_file)) {
    phecode_df <- phecode_file
  } else {
    cli::cli_abort("{.arg phecode_file} must be a file path or a dataframe")
  }

  # Format covariate data
  if (is.character(covariate_files)) {
    covariate_df <- pmbb_format_covariates(covariate_files, {{ covariate_cols }}, !!ensym(covariate_population_col), populations)
  } else if (is.data.frame(covariate_files)) {
    covariate_df <- covariate_files
  } else {
    cli::cli_abort("{.arg covariate_files} must be file paths or a dataframe")
  }

  # Run PheWAS
  cli::cli_progress_step("Running PheWAS")

  phewas_combos <- tidyr::expand_grid(genotype_masks = mask_genotypes_list, population = populations) %>%
    dplyr::mutate(mask_name = purrr::imap(genotype_masks, ~.y)) %>%
    tidyr::unnest(mask_name)

  phewas_res_list <- purrr::pmap(phewas_combos, \(genotype_masks, population, mask_name) {
    if (population != "ALL") {
      covariate_df <- covariate_df %>%
        dplyr::filter(!!ensym(covariate_population_col) == population) %>%
        dplyr::select(-!!ensym(covariate_population_col))
    } else {
      covariate_df <- covariate_df %>%
        dplyr::select(-!!ensym(covariate_population_col))
    }

    allele_counts <- genotype_masks %>%
      purrr::pluck("genotypes") %>%
      count(genotype)

    additive <- dplyr::if_else(genotype_masks %>% purrr::pluck("mask_type") == "single", TRUE, FALSE)

    phewas_res <- PheWAS::phewas(
      genotypes = genotype_masks %>% purrr::pluck("genotypes"),
      phenotypes = phecode_df,
      covariates = covariate_df,
      # additive.genotypes = additive,
      ...
    )

    dplyr::tibble(
      mask_name = mask_name,
      population = population,
      allele_counts = list(allele_counts),
      covariates = list(names(covariate_df %>% dplyr::select(-PMBB_ID))),
      phewas_res
    )
  }) %>%
    purrr::list_rbind()

  return(phewas_res_list)
}
